File: src/main/ts/tinydiff/DiffImpl.ts
namespace tinydiff {

  interface IMatch {
    x : number;
    length : number;
  }

  /**
   * createDiffImpl
   * @author Kazuhiko Arase
   */
  export function createDiffImpl() : IDiff {

    var _target : IDiffComparable;

    var matches : IMatch[];

    var snake = function(k : number, y : number) {

      var x = y - k;
      var length = 0;

      while(x < _target.getM() && y < _target.getN() &&
          _target.equals(x, y) ) {
        x += 1;
        y += 1;
        length += 1;
      }

      if (length > 0) {
        matches[y] = createMatch(x, length);
      }

      return y;
    };

    return {
      onp(target) {

        if (target.getM() > target.getN() ) {
          throw target.getM() + '>' + target.getN();
        }

        _target = target;

        matches = new Array(target.getN() + 1);
        matches[0] = createMatch(0, 0);

        var delta = target.getN() - target.getM();
        var offset = target.getM() + 1;

        var fp : number[] = new Array(target.getM() + target.getN() + 3);
        for (var i = 0; i < fp.length; i += 1) {
          fp[i] = -1;
        }

        var k : number;
        var p = -1;

        do {

          p += 1;

          for(k = -p; k < delta; k += 1) {
            fp[offset + k] = snake(k,
                Math.max(fp[offset + k - 1] + 1, fp[offset + k + 1] ) );
          }

          for(k = delta + p; k > delta; k -= 1) {
            fp[offset + k] = snake(k,
                Math.max(fp[offset + k - 1] + 1, fp[offset + k + 1] ) );
          }

          // k == delta
          fp[offset + k] = snake(k,
              Math.max(fp[offset + k - 1] + 1, fp[offset + k + 1] ) );

        } while (fp[offset + k] != target.getN() );

        // returns edit distance
        return delta + 2 * p;
      },

      trace(tracer) {

        var x = _target.getM();
        var y = _target.getN();
        var lastX : number;
        var lastY : number;

        var match : IMatch = null;

        while (x > 0 || y > 0) {

          //-------------------------
          // b only

          lastX = x;
          lastY = y;
          while (y >= 0) {
             match = matches[y];
            if (match != null && match.x <= x) {
              // found.
              break;
            }
            y -= 1;
          }
          if (lastY != y) {
            tracer.trace(lastX - x, lastY - y);
          }

          //-------------------------
          // a only

          lastX = x;
          lastY = y;
          while (x > match.x) {
            x -= 1;
          }
          if (lastX != x) {
            tracer.trace(lastX - x, lastY - y);
          }

          //-------------------------
          // match part

          lastX = x;
          lastY = y;
          while (x > match.x - match.length) {
            x -= 1;
            y -= 1;
          }
          if (lastX != x) {
            tracer.trace(lastX - x, lastY - y);
          }
        }
      }
    };
  }

  function createMatch(x : number, length : number) : IMatch {
    return {
      x : x,
      length : length,
      toString : () => 'Match(x=' + x + ',length=' + length + ')'
    };
  }
}

File: src/main/java/tinydiff/DiffImpl.java
package tinydiff;

/**
 * DiffImpl
 * @author Kazuhiko Arase
 */
class DiffImpl implements IDiff {

  private IDiffComparable target;

  private Match[] matches;

  public DiffImpl() {
  }

  public int onp(final IDiffComparable target) {

    if (target.getM() > target.getN() ) {
      throw new IllegalArgumentException(
          target.getM() + ">" + target.getN() );
    }

    this.target = target;

    matches = new Match[target.getN() + 1];
    matches[0] = new Match(0, 0);

    final int delta = target.getN() - target.getM();
    final int offset = target.getM() + 1;

    final int[] fp = new int[target.getM() + target.getN() + 3];
    for (int i = 0; i < fp.length; i += 1) {
      fp[i] = -1;
    }

    int k;
    int p = -1;

    do {

      p += 1;

      for(k = -p; k < delta; k += 1) {
        fp[offset + k] = snake(k,
            Math.max(fp[offset + k - 1] + 1, fp[offset + k + 1] ) );
      }

      for(k = delta + p; k > delta; k -= 1) {
        fp[offset + k] = snake(k,
            Math.max(fp[offset + k - 1] + 1, fp[offset + k + 1] ) );
      }

      // k == delta
      fp[offset + k] = snake(k,
          Math.max(fp[offset + k - 1] + 1, fp[offset + k + 1] ) );

    } while (fp[offset + k] != target.getN() );

    // returns edit distance
    return delta + 2 * p;
  }

  private int snake(int k, int y) {

    int x = y - k;
    int length = 0;

    while(x < target.getM() && y < target.getN() && target.equals(x, y) ) {
      x += 1;
      y += 1;
      length += 1;
    }

    if (length > 0) {
      matches[y] = new Match(x, length);
    }

    return y;
  }

  public void trace(final IPathTracer tracer) {

    int x = target.getM();
    int y = target.getN();
    int lastX;
    int lastY;

    Match match = null;

    while (x > 0 || y > 0) {

      //-------------------------
      // b only

      lastX = x;
      lastY = y;
      while (y >= 0) {
         match = matches[y];
        if (match != null && match.x <= x) {
          // found.
          break;
        }
        y -= 1;
      }  
      if (lastY != y) {
        tracer.trace(lastX - x, lastY - y);
      }

      //-------------------------
      // a only

      lastX = x;
      lastY = y;
      while (x > match.x) {
        x -= 1;
      }
      if (lastX != x) {
        tracer.trace(lastX - x, lastY - y);
      }

      //-------------------------
      // match part

      lastX = x;
      lastY = y;
      while (x > match.x - match.length) {
        x -= 1;
        y -= 1;
      }
      if (lastX != x) {
        tracer.trace(lastX - x, lastY - y);
      }
    }
  }

  private static final class Match {
    public final int x;
    public final int length;
    public Match(final int x, final int length) {
      this.x = x;
      this.length = length;
    }
    @Override
    public String toString() {
      return "Match(x=" + x + ",length=" + length + ")";
    }
  }
}
